backends/nxp/runtime/NeutronBackend.cpp

```diff
@@ -0,0 +1,411 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Implementation of the backend for the NXP Neutron NPU.
+ */
+
+#include <executorch/runtime/backend/interface.h>
+#include <executorch/runtime/core/error.h>
+#include <executorch/runtime/core/evalue.h>
+
+#include "NeutronDriver.h"
+#include "NeutronErrors.h"
+
+using namespace std;
+
+namespace torch {
+namespace executor {
+
+// All the memory need to be aligned with 16
+#define BUFFER_ALIGNMENT 16
+#define ALIGN_SIZE(size) \
+  ((size + BUFFER_ALIGNMENT - 1) & (~(BUFFER_ALIGNMENT - 1)))
+
+/* Header schema:
+        +----------------------------------+-----------------------------------+
+        | Input TensorFormats length (1B)  | Output TensorFormats length (1B)  |
+        +----------------------------------+-----------------------------------+
+        | 1st input tensor format (1B)     | [nth* input tensor format (1B)]   |
+        +----------------------------------+-----------------------------------+
+        | 1st output tensor format (1B)    | [nth* output tensor format (1B)]  |
+        +----------------------------------+-----------------------------------+
+*/
+#define ITEM_SIZE 1 // 1 Byte
+#define INPUT_TENSOR_FORMAT_LEN_POS 0
+#define OUTPUT_TENSOR_FORMAT_LEN_POS 1
+#define INPUT_TENSOR_FORMAT_ARRAY_ADDR(base) (base + 2 * ITEM_SIZE)
+#define OUTPUT_TENSOR_FORMAT_ARRAY_ADDR(base) \
+  (base + 2 * ITEM_SIZE + base[INPUT_TENSOR_FORMAT_LEN_POS])
+#define PAYLOAD_ADDR(base)                                 \
+  (base +                                                  \
+   ALIGN_SIZE(                                             \
+       2 * ITEM_SIZE + base[INPUT_TENSOR_FORMAT_LEN_POS] + \
+       base[OUTPUT_TENSOR_FORMAT_LEN_POS]))
+
+// Aggregate neutron model handle and data structures into one.
+typedef struct {
+  int numInputs = 0;
+  int numOutputs = 0;
+  uint32_t scratchSize = 0;
+  NeutronModelConfig mcfg;
+  NeutronDataConfig dcfg;
+  NeutronModelHandle nmh = NULL;
+  const uint8_t* inputTranspositionFlags;
+  const uint8_t* outputTranspositionFlags;
+} NeutronConfig;
+
+// Applied on outputs.
+template <typename T>
+void transposeToChannelFirst(
```

Comment:
OK for now, but I wonder is there another way to do this, i.e. set the dim_order on the output and let the portable or someone else take care of this?

Reply:
From what @jirioc explained below, I think this could be removed by inserting a transpose node instead of setting a flag